In [2]:
import pandas as pd

In [3]:
titanic = pd.read_csv('train.csv')
t = titanic.set_index(['PassengerId'])
pd.set_option('display.max_rows', None)

In [4]:
result = pd.DataFrame(index=t.columns, columns=t.columns)

In [5]:
for outer_column in t.columns:
    ff = t[t[outer_column].duplicated(keep=False)][outer_column].value_counts()
    if outer_column in ['Name']:
        continue
    for column in t.columns:
        sum = 0
        if column in ['Name', outer_column]:
            continue
        for i in ff.items():
            percent = round(t[t[outer_column] == i[0]].fillna(-1)[column].value_counts(normalize=True)*100, 2).max()
            column_value = t[t[outer_column] == i[0]].fillna(-1)[column].value_counts().sort_values(ascending=False).head(1).index[0]
            sum += i[1] * percent
        result[outer_column][column] = round(sum / ff.sum(), 2)

result.dropna(axis='index', how='all').dropna(axis='columns', how='all')

Survived Pclass    Sex    Age  SibSp  Parch Ticket   Fare  Cabin  \
Survived      NaN   67.9  78.67  65.19   63.3  63.03  84.88  77.96  82.52   
Pclass      57.02    NaN  55.11  59.46  55.11  55.05  100.0  97.98  100.0   
Sex         78.68  64.76    NaN  66.19  65.77  67.76  72.38  75.69  74.76   
Age         19.86  19.87  19.87    NaN  20.42  20.56  47.38  35.01  48.54   
SibSp       68.24  68.24  68.24  68.34    NaN  71.01  84.01  86.15  94.17   
Parch        76.1   76.1  76.09  78.94  79.01    NaN  87.21  90.43   93.2   
Ticket       1.35    1.8   1.34  11.46   3.93   2.81    NaN  41.44  85.44   
Fare         6.06  11.22   6.29  16.76   9.54   6.96  99.71    NaN  89.32   
Cabin        77.1  77.11  77.11  74.21  77.11  77.08  91.57  88.04    NaN   
Embarked    72.28  72.28  72.28  79.51  72.28  72.25  99.42  97.23  97.09   

         Embarked  
Survived    63.78  
Pclass      57.37  
Sex         64.91  
Age         19.91  
SibSp       68.16  
Parch       76.04  
Ticket        1.8  
Fare          9.9  
Cabin       77.28  
Embarked      NaN

ایده این بود:
بر اساس هر ستون آمدم، آنهایی که تکرارشون بیش از یکبار هست را پیدا کردم
خروجی میشد لیستی از اون ستون که تعداد تکرارش از یکبار بیشتر هست.
مثلا بر اساس ستون تیکت اومدم اونهایی که تکرارشون بیش از یکبار هست را در آوردم:

In [6]:
ff = t[t['Ticket'].duplicated(keep=False)]['Ticket'].value_counts()
ff

Ticket
1601                7
CA. 2343            7
347082              7
3101295             6
347088              6
CA 2144             6
382652              5
S.O.C. 14879        5
4133                4
113760              4
PC 17757            4
2666                4
W./C. 6608          4
349909              4
113781              4
LINE                4
19950               4
347077              4
17421               4
347742              3
363291              3
239853              3
13502               3
230080              3
29106               3
345773              3
35273               3
C.A. 31921          3
371110              3
SC/Paris 2123       3
F.C.C. 13529        3
110413              3
PC 17582            3
PC 17572            3
C.A. 34651          3
24160               3
248727              3
PC 17760            3
PC 17755            3
110152              3
PC 17608            2
PC 17477            2
P/PP 3381           2
367226              2
2691                2
117

در گام بعدی گفتم من میخوام از بین اونهایی که تکراری هستند 
برای مثال در اینجا تیکت 1601 که اومده و هفت بار تکرار شده، من میخوام 
توزیع این هفت عنصر را روی تمامی ستونهای دیگه نگاه کنم.
(یعنی حالا که هفت تا تیکت تکراری دارم، وضعیت این هفت تا تیکت تکراری تو ستونهای دیگه نسبت به هم چطوری هست)

In [7]:
print(f'Total Ticket {ff.index[0]} is {ff.values[0]}')

for column in t.columns:
    if column in ['Name', 'Ticket']:
        continue
    for i in ff.items():
        percent = round(t[t['Ticket'] == i[0]].fillna(-1)[column].value_counts(normalize=True)*100, 2)#.max()
        column_value = t[t['Ticket'] == i[0]].fillna(-1)[column].value_counts().sort_values(ascending=False)#.head(1).index[0]
        print(column_value, percent)
        print("===============================")
        break

Total Ticket 1601 is 7
Survived
1    5
0    2
Name: count, dtype: int64 Survived
1    71.43
0    28.57
Name: proportion, dtype: float64
Pclass
3    7
Name: count, dtype: int64 Pclass
3    100.0
Name: proportion, dtype: float64
Sex
male    7
Name: count, dtype: int64 Sex
male    100.0
Name: proportion, dtype: float64
Age
-1.0     3
 32.0    2
 28.0    1
 26.0    1
Name: count, dtype: int64 Age
-1.0     42.86
 32.0    28.57
 28.0    14.29
 26.0    14.29
Name: proportion, dtype: float64
SibSp
0    7
Name: count, dtype: int64 SibSp
0    100.0
Name: proportion, dtype: float64
Parch
0    7
Name: count, dtype: int64 Parch
0    100.0
Name: proportion, dtype: float64
Fare
56.4958    7
Name: count, dtype: int64 Fare
56.4958    100.0
Name: proportion, dtype: float64
Cabin
-1    7
Name: count, dtype: int64 Cabin
-1    100.0
Name: proportion, dtype: float64
Embarked
S    7
Name: count, dtype: int64 Embarked
S    100.0
Name: proportion, dtype: float64


از مقادیری که در بالا هست من بیشترین درصد را به همراه لیبلی که بیشترین توزیع را ثبت کرده به عنوان نتیجه‌ی نهایی خارج میکنم
برای مثال برای تیکت 1601 وقتی توزیع روی سروایود را نگاه کنیم، 
نزدیک به هفتاد درصد از تیکتهایی که  شماره 1601 را دارند، مقدار سروایود آنها یک بوده است
پس من بالاترین درصد گزارش شده و مقدار لیبل آن که اینجا یک هست را به عنوان توزیع سروایود ها روی تیکت ۱۶۰۱ در نظر میگیرم.

همین کار را برای تمامی تیکت های دیگر که بیش از یکبار تکرار  شده اند انجام خواهم داد.

توجه شود که نه تنها توزیع سروایود بلکه توزیع روی تمامی ستونها را بررسی میکنم که در زیر آمده است.

In [8]:
ff = t[t['Ticket'].duplicated(keep=False)]['Ticket'].value_counts()

for column in t.columns:
    if column in ['Name', 'Ticket']:
        continue
    for i in ff.items():
        percent = round(t[t['Ticket'] == i[0]].fillna(-1)[column].value_counts(normalize=True)*100, 2).max()
        column_value = t[t['Ticket'] == i[0]].fillna(-1)[column].value_counts().sort_values(ascending=False).head(1).index[0]
        print(f'Ticket: {i[0]}, duplicate in {i[1]} times and {percent}% tell {column} is {column_value}')
    print("----------------------------------------------------------")   


Ticket: 1601, duplicate in 7 times and 71.43% tell Survived is 1
Ticket: CA. 2343, duplicate in 7 times and 100.0% tell Survived is 0
Ticket: 347082, duplicate in 7 times and 100.0% tell Survived is 0
Ticket: 3101295, duplicate in 6 times and 100.0% tell Survived is 0
Ticket: 347088, duplicate in 6 times and 100.0% tell Survived is 0
Ticket: CA 2144, duplicate in 6 times and 100.0% tell Survived is 0
Ticket: 382652, duplicate in 5 times and 100.0% tell Survived is 0
Ticket: S.O.C. 14879, duplicate in 5 times and 100.0% tell Survived is 0
Ticket: 4133, duplicate in 4 times and 100.0% tell Survived is 0
Ticket: 113760, duplicate in 4 times and 100.0% tell Survived is 1
Ticket: PC 17757, duplicate in 4 times and 75.0% tell Survived is 1
Ticket: 2666, duplicate in 4 times and 100.0% tell Survived is 1
Ticket: W./C. 6608, duplicate in 4 times and 100.0% tell Survived is 0
Ticket: 349909, duplicate in 4 times and 100.0% tell Survived is 0
Ticket: 113781, duplicate in 4 times and 50.0% tell S

آخرین کاری که انجام میدهم، این هست که میانگین وزن دار توزیع ها را حساب میکنم.
در واقع براساس تیکت هایی که بیش از یکبار تکرار شده بودند
من توزیع این تیکت ها را روی تمامی ستونها بدست آوردم

ولی یک سوال اساسی باید برای من حل شود و آن این است که :
بالاخره برای مثال توزیع ستون سروایود روی تیکت چگونه هست؟

برای این کار انتخاب من میانگین وزندار هست
یعنی من برای مثال فهمیدم که تیکت 1601 هفت بار تکرار شده و توزیع آن هم روی سرواید حدود هفتاد درصد میباشد.
همین را برای تیکت
'CA. 2343'
هم فهمیدم که آن هم مثلا هفت بار تکرار شده و توزیع آن صد درصد هست، یعنی همشون سروایود یکسانی دارند 

من در نهایت برای اینکه بالاخره بگم توزیع نهایی چطوری هست، از میانگین وزندار استفاده میکنم
و بین تمامی تیکت هایی که بیش از یکبار تکرار شدند، بر اساس تعداد تکرارشون و بالاترین درصدی که گزارش کردند، میانگین وزندار میگیرم،

و در نهایت میتوانم بگویم میانگین وزندار توزیع سروایودها روی تیکتهای تکراری به چه صورت هست.
و این کار را برای تمامی ستونهای دیگه به غیر از سروایود هم انجام میدم
و در واقع میانگین وزندار توزیع تمامی ستونها را روی تیکت های تکراری خواهم دید

In [9]:
ff = t[t['Ticket'].duplicated(keep=False)]['Ticket'].value_counts()

for column in t.columns:
    sum = 0
    if column in ['Name', 'Ticket']:
        continue
    for i in ff.items():
        percent = round(t[t['Ticket'] == i[0]].fillna(-1)[column].value_counts(normalize=True)*100, 2).max()
        column_value = t[t['Ticket'] == i[0]].fillna(-1)[column].value_counts().sort_values(ascending=False).head(1).index[0]
        sum += i[1] * percent

    print(f'{column}:    \t', sum/ff.sum())
    print("----------------------------------------------------------")   

Survived:    	 84.88409883720932
----------------------------------------------------------
Pclass:    	 100.0
----------------------------------------------------------
Sex:    	 72.3841569767442
----------------------------------------------------------
Age:    	 47.38348837209301
----------------------------------------------------------
SibSp:    	 84.01194767441861
----------------------------------------------------------
Parch:    	 87.2096511627907
----------------------------------------------------------
Fare:    	 99.70930232558139
----------------------------------------------------------
Cabin:    	 91.56985465116279
----------------------------------------------------------
Embarked:    	 99.41863372093025
----------------------------------------------------------


همین روند فوق که برای ستون تیکت رفتم را برای تمامی ستونهای دیگر هم میرم در واقع جدولی که در ادامه آمده است نتیجه‌ی همان هست
که در این جا ستونهایش به ترتیب نقش همین تیکت که تاکنون صحبت شد را بازی میکنند.
یعنی در هر ستون، من به دنبال آنهایی رفتم که تعداد تکرارشان بیش از یکبار هست 
و سپس میانگین وزندار توزیع تمامی ردیفهای این جدول را روی این ستون گفته شده زده یادداشت کرده ام
اگر ستون تیکت را مشاهده کنید، دقیقا خروجی بخش بالا را مشاهده خواهید کرد:

In [11]:
for outer_column in t.columns:
    ff = t[t[outer_column].duplicated(keep=False)][outer_column].value_counts()
    if outer_column in ['Name']:
        continue
    for column in t.columns:
        sum = 0
        if column in ['Name', outer_column]:
            continue
        for i in ff.items():
            percent = round(t[t[outer_column] == i[0]].fillna(-1)[column].value_counts(normalize=True)*100, 2).max()
            column_value = t[t[outer_column] == i[0]].fillna(-1)[column].value_counts().sort_values(ascending=False).head(1).index[0]
            sum += i[1] * percent
        result[outer_column][column] = round(sum / ff.sum(), 2)

result.dropna(axis='index', how='all').dropna(axis='columns', how='all')

Survived Pclass    Sex    Age  SibSp  Parch Ticket   Fare  Cabin  \
Survived      NaN   67.9  78.67  65.19   63.3  63.03  84.88  77.96  82.52   
Pclass      57.02    NaN  55.11  59.46  55.11  55.05  100.0  97.98  100.0   
Sex         78.68  64.76    NaN  66.19  65.77  67.76  72.38  75.69  74.76   
Age         19.86  19.87  19.87    NaN  20.42  20.56  47.38  35.01  48.54   
SibSp       68.24  68.24  68.24  68.34    NaN  71.01  84.01  86.15  94.17   
Parch        76.1   76.1  76.09  78.94  79.01    NaN  87.21  90.43   93.2   
Ticket       1.35    1.8   1.34  11.46   3.93   2.81    NaN  41.44  85.44   
Fare         6.06  11.22   6.29  16.76   9.54   6.96  99.71    NaN  89.32   
Cabin        77.1  77.11  77.11  74.21  77.11  77.08  91.57  88.04    NaN   
Embarked    72.28  72.28  72.28  79.51  72.28  72.25  99.42  97.23  97.09   

         Embarked  
Survived    63.78  
Pclass      57.37  
Sex         64.91  
Age         19.91  
SibSp       68.16  
Parch       76.04  
Ticket        1.8  
Fare          9.9  
Cabin       77.28  
Embarked      NaN